Write log with level :error when an exception got raised when trying to deliver an email.

ignatiusreza opened the pull request:
Before:

>  "Sent mail to [recipients]" got written even when we fail to deliver.

After:

>  "[exception_class] was raised when sending mail to [recipients]" will be written

This changes helps in understanding which emails got delivered and which emails does not when parsing hundreds of lines of mailer logs. Current behavior is misleading, because "Sent mail to" gives of the sense that emails got successfully sent, even when it's not.

egilburg commented on actionmailer/lib/action_mailer/log_subscriber.rb:
```diff
@@ -6,9 +6,16 @@ module ActionMailer
-        "\nSent mail to #{recipients} (#{event.duration.round(1)}ms)"
+      if exception = event.payload[:exception]
+        error do
+          recipients = Array(event.payload[:to]).join(', ')
```
this line can be extracted out of the `if` statement.

ignatiusreza replied:
there're several other places where `"#{event.duration.round(1)}ms"` shows up.. how about we make a private method like:

```
def duration(event)
  "#{event.duration.round(1)}ms"
end
```

to further reduce duplication?

egilburg replied:
Sure, call it `duration_in_ms`

ignatiusreza replied:
done

egilburg commented on actionmailer/lib/action_mailer/log_subscriber.rb:
```diff
@@ -6,9 +6,16 @@ module ActionMailer
+
+      if exception = event.payload[:exception]
+        error do
+          "\n#{exception.first} was raised when sending mail to #{recipients} (#{duration_in_ms(event)})"
```
is `exception` always guaranteed to be an array? If not, you can wrap it in `Array()` similar to `recipients`.

ignatiusreza replied:
The comments on `ActiveSupport::Notifications` seems to confirm that

```
  # If an exception happens during that particular instrumentation the payload will
  # have a key <tt>:exception</tt> with an array of two elements as value: a string with
  # the name of the exception class, and the exception message.
```

there's also a test for it..

ref:
- https://github.com/ignatiusreza/rails/blob/action_mailer_log_subscriber_on_exception/activesupport/lib/active_support/notifications.rb#L64-L66
- https://github.com/ignatiusreza/rails/blob/action_mailer_log_subscriber_on_exception/activesupport/test/notifications_test.rb#L209-L221

egilburg replied:
Thanks!

cirojr commented:
Hi! I'm suffering trying to send an email for the first time and getting none error logs to help me. Is this exception raise already implemented?

eileencodes commented on actionmailer/lib/action_mailer/log_subscriber.rb:
```diff
@@ -25,14 +30,25 @@ def process(event)
     end
+
+    private
+
```
✂️  Remove this space per the styleguide.

eileencodes commented on actionmailer/test/log_subscriber_test.rb:
```diff
@@ -35,6 +41,29 @@ def test_deliver_is_notified
+    assert_raise Net::SMTPAuthenticationError do
+      mail.deliver_now
+    end
+    wait
```
I'm not sure if we have an explict style guide for this, but tests are easier to read when it goes assignments, assertions, with spaces inbetween so I think this would read better as:

```ruby
mail = BaseMailer.welcome

assert_raise Net::SMTPAuthenticationError do
  mail.deliver_now
end

wait
```

eileencodes commented on actionmailer/test/log_subscriber_test.rb:
```diff
@@ -35,6 +41,29 @@ def test_deliver_is_notified
+    assert_equal(0, @logger.logged(:info).size)
+
+    assert_equal(1, @logger.logged(:error).size)
+    assert_match(/Net::SMTPAuthenticationError was raised when sending mail to system@test.lindsaar.net/, @logger.logged(:error).first)
```
Can this use assert_equal instead of assert_match?

eileencodes replied:
Ah I see you're looking at `@logger.logged` so you need match 👍

cirojr replied:
Thanks for the contribution Eileen! It needs to be approved by someone to be merged in the master branch of rails?
To use it now I would have to fork rails in my machine and copy this code, right?
I want to be able to contribute too. I've read a post of yours about it. Any other advice on it would be very appreciated! :D

eileencodes replied:
Hey @cirojr it's really awesome that you want to contribute. Regarding using this code; before it's merged into Rails, you could fork and cherrypick this commit, or you could link right to the branch like:

`gem 'rails', github: 'gnatiusreza/rails', branch: 'action_mailer_log_subscriber_on_exception'

Since everyone who is on this PR gets an email every time you comment this isn't a good place to discuss contributing. You can DM me on twitter (same handle here @eileencodes) if you want to chat more about contributing. I'd be happy to point you in the right direction. 😄

ignatiusreza commented:
updated per review, thanks @eileencodes !

ignatiusreza committed:
Write log with level :error when an exception got raised
when trying to deliver an email.

Before:

  "Sent mail to [recipients]" got written even when we fail to deliver.

After:

  "[exception_class] was raised when sending mail to [recipients]" will be written

This changes helps in understanding which emails got delivered and which
emails does not when parsing hundreds of lines of mailer logs.

ignatiusreza commented:
Hi guys, sorry for re-raising an old PR.. one of my team mates was just got bitten by this issue again.. would really love to get it fixed.. is that anything I can do to help move this forward?

simi commented:
:information_source: error during delivery is still logged as success. IMHO it will make sense to log errors done during delivery. @eileencodes would you mind to take a look again? Is there any blocker for this PR?
